File: code/backend/internal/login/service.go
package login

import (
	"errors"
	"log"
)

type AuthService interface {
	Authenticate(username, password string) (string, error)
}

type authService struct {
	repo UserRepository
}

func NewAuthService(repo UserRepository) AuthService {
	return &authService{repo: repo}
}

func (s *authService) Authenticate(username, password string) (string, error) {
	// サービス層での認証処理開始ログ
	log.Printf("[Service] Authenticate called. Username: %s", username)
	user, err := s.repo.FindByUsername(username)
	if err != nil {
		log.Printf("[Service] ユーザー検索エラー: %v", err)
		return "", err
	}
	if user == nil || user.Password != password {
		log.Printf("[Service] 認証失敗: %s", username)
		return "", errors.New("invalid credentials")
	}
	log.Printf("[Service] 認証成功: %s", username)
	// 本来はここでJWTなどを返すが、サンプルなので固定トークン
	return "dummy-token", nil
}

File: code/backend/src/features/login/service.rs
use crate::features::login::repository::{UserRepository, UserRecord};
use anyhow::{Result, anyhow};

pub struct LoginService<R: UserRepository> {
    pub repo: R,
}

impl<R: UserRepository> LoginService<R> {
    pub fn new(repo: R) -> Self {
        Self { repo }
    }

    pub fn authenticate(&self, username: &str, password: &str) -> Result<UserRecord> {
        println!("[SERVICE] authenticate: username={}", username);
        let user = self.repo.find_by_username(username)?;
        if let Some(user) = user {
            println!("[SERVICE] ユーザー発見: username={}", user.username);
            if user.password == password {
                println!("[SERVICE] パスワード一致: username={}", user.username);
                Ok(user)
            } else {
                println!("[SERVICE] パスワード不一致: username={}", user.username);
                Err(anyhow!("Invalid password"))
            }
        } else {
            println!("[SERVICE] ユーザー未発見: username={}", username);
            Err(anyhow!("User not found"))
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    struct MockRepo;
    impl UserRepository for MockRepo {
        fn find_by_username(&self, username: &str) -> Result<Option<UserRecord>> {
            if username == "alice" {
                Ok(Some(UserRecord {
                    id: 1,
                    username: "alice".to_string(),
                    password: "alicepw".to_string(),
                    display_name: "Alice".to_string(),
                }))
            } else {
                Ok(None)
            }
        }
    }

    #[test]
    fn test_authenticate_success() {
        let service = LoginService::new(MockRepo);
        let user = service.authenticate("alice", "alicepw").unwrap();
        assert_eq!(user.username, "alice");
    }

    #[test]
    fn test_authenticate_wrong_password() {
        let service = LoginService::new(MockRepo);
        assert!(service.authenticate("alice", "wrongpw").is_err());
    }

    #[test]
    fn test_authenticate_user_not_found() {
        let service = LoginService::new(MockRepo);
        assert!(service.authenticate("bob", "bobpw").is_err());
    }
}
